jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/testimpl/WebSocketHandler.java

```diff
@@ -0,0 +1,457 @@
+            WebSocketRemoteEndpointImpl remoteEndpoint = new WebSocketRemoteEndpointImpl(connection);
+            remoteEndpoint.open(); // TODO why is this needed?
+
+            // TODO abstract the creation of a local Endpoint
+            WebSocketLocalEndpoint localEndpoint = new WebSocketLocalEndpoint()
+            {
+                @Override
+                public Logger getLog()
+                {
+                    return LOG;
+                }
+
+                @Override
+                public boolean isOpen()
+                {
+                    return connection.isOpen();
+                }
+
+                @Override
+                public void onOpen()
+                {
+                    LOG.debug("onOpen {}",this);
+                    TextFrame text = new TextFrame();
+                    text.setPayload("Opened!");
+
+                    // TODO do I need to mask this myself? if so then why?
+                    remoteEndpoint.sendFrame(text, new Callback()
+                    {
+                        @Override
+                        public void succeeded()
+                        {
+                            LOG.debug("onOpen write!");
+                        }
+
+                        @Override
+                        public void failed(Throwable x)
+                        {
+                            LOG.warn(x);
+                        }
+                    });
+                }
+
+                @Override
+                public void onText(Frame frame, Callback callback)
+                {
+                    ByteBuffer payload = frame.getPayload();
+                    // TODO why does the session have to demask?
+                    if (frame.isMasked())
+                    {
+                        ParserDeMasker demask = new ParserDeMasker(); // TODO recycle
+                        demask.reset(frame);
+                        demask.process(payload);
+                        // TODO this doesn't appear to work???
+                    }
+                    String text = BufferUtil.toUTF8String(payload);
+
+                    LOG.debug("onText {} / {}",text,frame);
+                    // TODO why not constructor injection on TextFrame?
+                    TextFrame textFrame = new TextFrame();
+
+                    // TODO why not String getter or ByteBuffer setter?
+                    textFrame.setPayload("echo: "+text);
+                    remoteEndpoint.sendFrame(textFrame, callback);
+                }
+            };
+
+            WebSocketCoreSession<ContainerLifeCycle,WebSocketCoreConnection,WebSocketLocalEndpoint,WebSocketRemoteEndpoint> session =
+                    new WebSocketCoreSession(this,connection)
+                    {
+                        // TODO why is core session abstract?
+                    };
+
+            //TODO Why can't this be done in constructor?
+            //TODO Why does session need the endpoint object?
```

Comment:
Order of creation is different in the different APIs and modes (server vs client).

Reply:
Ah OK - need to test client side as well then to get a feel for that.  yuck I hate clients!